src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -248,6 +245,10 @@ private void CheckInvariants()
                 return this;
             }
 
+            // If we were given a new state map, then our IDs must correspond to it.  Otherwise, we just
+            // pass our existing ids along as nothing changed with them.
+            var projectIds = idToProjectStateMap == _projectIdToProjectStateMap ? this.ProjectIds : idToProjectStateMap.Keys.ToImmutableArray();
```

Comment:
I think the "idea" behind the old pattern was enumerating keys and producing the new ImmutableArray wasn't cheap, and the idea was we're going to reuse the list from snapshot to snapshot and not even reallocate unless it changes. With this change, we're likely going to rebuild it any time a project changes, even if a project isn't added or removed. I'm guessing the vast majority of the calls to this has at least one change in the state map, so we'll probably be reallocating the keys every single time.

Now: does that matter? Only PerfView can say. I'd say do a few minutes of Roslyn dogfooding and see how hot this is, and if it's a problem then continue to adjust. I'm reminded of https://github.com/dotnet/roslyn/issues/53769 where some more expensive work in Branch here actually showed up in traces and we fixed it. That was creating hash sets was definitely more expensive than the code here.

One way to think about it: this change immediately means that opening a large solution is more O(n^2) than before, since every mutation is now computing this list again and again, which itself is an O(n) operation.

Reply:
ok.  i'll close this out.